usr/src/uts/common/fs/zfs/zil.c

```diff
@@ -1208,22 +1208,17 @@ zil_lwb_write_issue(zilog_t *zilog, lwb_t *lwb)
 	tx = dmu_tx_create(zilog->zl_os);
 
 	/*
-	 * Since we are not going to create any new dirty data and we can even
-	 * help with clearing the existing dirty data, we should not be subject
-	 * to the dirty data based delays.
-	 * We (ab)use TXG_WAITED to bypass the delay mechanism.
-	 * One side effect from using TXG_WAITED is that dmu_tx_assign() can
-	 * fail if the pool is suspended.  Those are dramatic circumstances,
-	 * so we return NULL to signal that the normal ZIL processing is not
-	 * possible and txg_wait_synced() should be used to ensure that the data
-	 * is on disk.
+	 * Since we are not going to create any new dirty data and we
+	 * can even help with clearing the existing dirty data, we
+	 * should not be subject to the dirty data based delays. We
+	 * (ab)use TXG_WAITED to bypass the delay mechanism.
 	 */
-	error = dmu_tx_assign(tx, TXG_WAITED);
-	if (error != 0) {
-		ASSERT3S(error, ==, EIO);
-		dmu_tx_abort(tx);
-		return (NULL);
+	while ((error = dmu_tx_assign(tx, TXG_WAITED)) != 0) {
+		VERIFY3S(error, ==, ERESTART);
```

Comment:
I understand that a `EIO` here would cause some problems.
But it is a very valid code that can be returned when `failmode` is `continue` and the pool gets suspended.
Simply panicing with such configuration seems like a too drastic response.
Perhaps this code should treat that kind of `EIO` same as it treats `ERESTART`.

Reply:
That seems like a reasonable approach. I'll try that out.